include/cudaq/Optimizer/Dialect/Quake/QuakeOps.td

```diff
@@ -91,46 +94,58 @@ def quake_AllocaOp : QuakeOp<"alloca", [MemoryEffects<[MemAlloc, MemWrite]>]> {
 def quake_ConcatOp : QuakeOp<"concat", [Pure]> {
   let summary = "Construct a qvec from a list of other qref/qvec values.";
   let description = [{
-    The `quake.concat` operation allows one to group a list of values of type
-    QRef and/or QVec into a new QVec aggregate.
-    ```
-    %veq = quake.concat %r1, %v1, %r2 : (!quake.qref, !quake.qvec<?>, !quake.qref) -> !quake.qvec<?>
+    The `concat` operation allows one to concatenate a list of SSA-values of
+    either type QRef or QVec into a new QVec vector.
+
+    Example:
+    ```mlir
+      %veq = quake.concat %r1, %v1, %r2 : (!quake.qref, !quake.qvec<?>,
+                                           !quake.qref) -> !quake.qvec<?>
     ```
   }];
 
-  let arguments = (ins Variadic<AnyQType>:$qbits);
+  let arguments = (ins Variadic<AnyQRefType>:$qbits);
   let results = (outs QVecType);
 
   let assemblyFormat = [{
     $qbits attr-dict `:` functional-type(operands, results)
   }];
 }
 
-// A ComputeActionOp is marked as having memory effects to prevent the MLIR
-// optimizer from eliding it as dead code. It will be transformed into a series
-// of CallOps.
-def quake_ComputeActionOp : QuakeOp<"compute_action",
-    [MemoryEffects<[MemRead, MemWrite]>]> {
+// A ComputeActionOp will be transformed into a series of CallOps.
+def quake_ComputeActionOp : QuakeOp<"compute_action"> {
   let summary = "Captures the compute/action/uncompute high-level idiom.";
   let description = [{
-    CUDA Quantum supports the high-level compute/action/uncompute idiom by
+    CUDA Quantum supports the high-level compute, action, uncompute idiom by
     providing a custom template function (class) that takes pure kernels (a
     callable like a λ) as arguments. This operation captures uses of the idiom
     and can be systematically expanded into a quantum circuit via successive
     transformations.
+
+    The `is_dagger` attribute can be used to "reverse" this idiom to one of
+    uncompute, action, compute.
+
+    The uncompute step is generated automatically by generating the adjoint of
+    the compute kernel.
   }];
 
-  // FIXME: The compute and action can be other sorts of Callable instances and
-  // need not be inline λ expressions.
   let arguments = (ins
     UnitAttr:$is_dagger,
     cc_LambdaType:$compute,
     cc_LambdaType:$action
   );
-  
+
+  let assemblyFormat = [{
+    (`<` `dag` $is_dagger^ `>`)? $compute `,` $action `:`
+      qualified(type(operands)) attr-dict
+  }];
+}
+
+def quake_ConvertControlOp : QuakeOp<"convert_ctrl"> {
```

Comment:
should you add a summary / description for this op?

Reply:
It is definitely needed if we decide to keep this as part of the IR long term. I'm still on the fence with this idea, so this is just a placeholder until a final consensus is reached.